plugins/usbdmx/USBDMXCom.cpp

```diff
@@ -0,0 +1,272 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * USBDMXCom.cpp
+ * The synchronous and asynchronous USBDMXCom widgets.
+ * Copyright (C) 2021 Peter Newman
+ */
+
+#include "plugins/usbdmx/USBDMXCom.h"
+
+#include <string.h>
+#include <string>
+
+#include "libs/usb/LibUsbAdaptor.h"
+#include "ola/Constants.h"
+#include "ola/Logging.h"
+#include "ola/StringUtils.h"
+#include "ola/util/Utils.h"
+#include "plugins/usbdmx/AsyncUsbSender.h"
+#include "plugins/usbdmx/ThreadedUsbSender.h"
+
+namespace ola {
+namespace plugin {
+namespace usbdmx {
+
+using std::string;
+using ola::usb::LibUsbAdaptor;
+
+namespace {
+
+// Why is this so long?
+static const unsigned int URB_TIMEOUT_MS = 500;
+static const unsigned char ENDPOINT = 0x02;
+enum { USBDMXCOM_MAX_FRAME_SIZE = (512*3)+5 };
+
+/*
+ * Create a USBDMX.com message to match the supplied DmxBuffer.
+ */
+size_t CreateFrame(const DmxBuffer &buffer, uint8_t frame[USBDMXCOM_MAX_FRAME_SIZE]) {
+  static uint8_t old_Values[512];
```

Comment:
Have a look at #1746 , you can just copy the buffer rather than writing all the values into another array just for comparison purposes.

Reply:
Where do i need to look at?
I made this this way, because no need to make a full copy if only one value changed, and i still have to compare them all..

Reply:
See OpenDeckWidget::SendDMX and it's internal_buffer:
https://github.com/OpenLightingProject/ola/pull/1746/files#diff-6ef0f8bbea30618491c7d8a882f9f77665c1c915218980ef16d13ffd59022c2bR46

> Where do i need to look at? I made this this way, because no need to make a full copy if only one value changed

OLA does some magic called copy on write so it sometimes won't even make a copy!

> i still have to compare them all..

You could do an equality check of the buffer, which does a nice efficient memcmp to decide if there is even anything to update.

Reply:
I don't really want to merge a PR which doesn't work or potentially doesn't, but I can try and help you with some of the changes within this PR.

But also a bit of this:
https://www.openlighting.org/ola/get-help/ola-faq/#Can_you_add_support_for_the_ltinsert-name-heregt_device